lib/output/writer/cassandra.go

```diff
@@ -0,0 +1,253 @@
+	mQueryLatency metrics.StatTimer
+	connLock      sync.RWMutex
+}
+
+// NewCassandra creates a new Cassandra writer type.
+func NewCassandra(conf CassandraConfig, log log.Modular, stats metrics.Type) (*Cassandra, error) {
+	c := Cassandra{
+		log:           log,
+		stats:         stats,
+		conf:          conf,
+		query:         fmt.Sprintf("INSERT INTO %s.%s JSON ?", conf.Keyspace, conf.Table),
+		mQueryLatency: stats.GetTimer("query.latency"),
+	}
+	return &c, nil
+}
+
+// Connect establishes a connection to Cassandra by calling ConnectWithContext and passing
+// a simple Background context.
+func (c *Cassandra) Connect() error {
+	return c.ConnectWithContext(context.Background())
+}
+
+// ConnectWithContext establishes a connection to Cassandra.
+func (c *Cassandra) ConnectWithContext(ctx context.Context) error {
+	c.connLock.Lock()
+	defer c.connLock.Unlock()
+	if c.session != nil {
+		return nil
+	}
+
+	var err error
+	conn := gocql.NewCluster(c.conf.Nodes...)
+	if c.conf.PasswordAuthenticator.Enabled {
+		conn.Authenticator = gocql.PasswordAuthenticator{
+			Username: c.conf.PasswordAuthenticator.Username,
+			Password: c.conf.PasswordAuthenticator.Password,
+		}
+	}
+	if conn.Consistency, err = gocql.ParseConsistencyWrapper(c.conf.Consistency); err != nil {
+		return fmt.Errorf("parsing consistency: %w", err)
+	}
+
+	min, err := time.ParseDuration(c.conf.Config.Backoff.InitialInterval)
+	if err != nil {
+		return fmt.Errorf("parsing backoff initial interval: %w", err)
+	}
+
+	max, err := time.ParseDuration(c.conf.Config.Backoff.MaxInterval)
+	if err != nil {
+		return fmt.Errorf("parsing backoff max interval: %w", err)
+	}
+
+	conn.RetryPolicy = &decorator{
+		NumRetries: int(c.conf.Config.MaxRetries),
+		Min:        min,
+		Max:        max,
+	}
+	session, err := conn.CreateSession()
+	if err != nil {
+		return fmt.Errorf("creating Cassandra session: %w", err)
+	}
+
+	c.session = session
+	c.log.Infof("Sending messages to Cassandra: %v\n", c.conf.Nodes)
+	return nil
+}
+
+// Connected returns true if the connection to Cassanrda is open.
+func (c *Cassandra) Connected() bool {
+	c.connLock.RLock()
+	session := c.session
+	c.connLock.RUnlock()
+
+	return session != nil && !session.Closed()
+}
+
+// Write writes a message to Cassandra by calling WriteWithContext with a simple Background context.
+func (c *Cassandra) Write(msg types.Message) error {
+	return c.WriteWithContext(context.Background(), msg)
+}
+
+// WriteWithContext writes a message to Cassandra.
+func (c *Cassandra) WriteWithContext(ctx context.Context, msg types.Message) error {
+	c.connLock.RLock()
+	session := c.session
+	c.connLock.RUnlock()
+
+	if c.session == nil {
+		return types.ErrNotConnected
+	}
+	if c.conf.AsyncParts {
+		var g errgroup.Group
+		msg.Iter(func(i int, p types.Part) error {
+			g.Go(func() error {
+				return c.writePart(session, p)
+			})
+			return nil
+		})
+		return g.Wait()
+	}
+
+	return msg.Iter(func(i int, p types.Part) error {
```

Comment:
We can probably speed things up here when there are >1 messages by using https://godoc.org/github.com/gocql/gocql#Batch